### PR TITLE
Tests II: more Julia function tests

### PR DESCRIPTION
**Pull request by jgreener64**
I added tests for more Julia functions, including some which have had Enzyme issues fixed recently. I also added a nested reverse mode test and turned on previous tests that now pass on older Julia versions.

All pass on my machine, let's see what CI says.

### CODE REVIEW

**Comment by wsmoses** [comment]
@jgreener64 can you isolate the 1.8 segfault into an issue?

**Comment by jgreener64** [comment]
Opened #970 and #971 and disabled the failing tests for now.

File: test/runtests.jl [hunk 1]
```diff
@@ -2286,8 +2417,8 @@ end
     @test autodiff(Reverse, f8, Active, Active(1.5))[1][1] == 0
     @test autodiff(Forward, f8, Duplicated(1.5, 1.0))[1]   == 0
 
-    # On Julia 1.6 the gradients are wrong (0.7 not 1.2) and on 1.7 it errors
-    @static if VERSION ≥ v"1.8-"
+    # On Julia 1.6 the gradients are wrong (0.7 not 1.2)
+    @static if VERSION ≥ v"1.7-"
```
Comment (wsmoses) [inline test/runtests.jl:2421]: Open a bug for this?

**Comment by wsmoses** [comment]
Can you isolate that 1.7 assertion error or segfault?

File: test/runtests.jl (hunk 1 above)
Comment (jgreener64) [reply test/runtests.jl:2421]: #973

**Comment by jgreener64** [comment]
I can't reproduce the Julia 1.7 issue locally, pushing some print statements to isolate it in CI.

**Comment by jgreener64** [comment]
Julia 1.7 issue reported at #974.

A Julia 1.6 issue in `try`/`catch` reported at #971.

File: test/runtests.jl (hunk 1 above)
Comment (wsmoses) [reply test/runtests.jl:2474]: remove the version now?

File: test/runtests.jl [hunk 2]
```diff
@@ -297,6 +297,153 @@ end
...
+
+    function f25(x)
+        try
+            sqrt(-1.0)
+            return 3x
+        catch
+            return 2x
+        end
+    end
+    @test autodiff(Reverse, f25, Active, Active(2.0))[1][1] == 2
```
Comment (wsmoses) [inline test/runtests.jl:401]: add the test_broken for 1.6?

Comment (jgreener64) [reply test/runtests.jl]: Done, unfortunately the `@test 1==1 broken=cond` syntax is not available on Julia 1.6 so we have to use an `if` and `@test_broken`.

By the way I can remove the print statements just before this is merged, they are for debugging.

**Comment by jgreener64** [comment]
Raised two new issues: https://github.com/EnzymeAD/Enzyme.jl/issues/1118 and https://github.com/EnzymeAD/Enzyme.jl/issues/1119.

**Comment by jgreener64** [comment]
I think the only unresolved issue is https://github.com/EnzymeAD/Enzyme.jl/issues/970, I could remove that test to merge this PR.

File: test/runtests.jl [hunk 3]
```diff
@@ -336,6 +336,148 @@ end
...
+    else
+        @test_broken autodiff(Reverse, f25, Active, Active(2.0))[1][1] == 2
+        @test_broken autodiff(Forward, f25, Duplicated(2.0, 1.0))[1]   == 2
+    end
+    println("Done 25")
+
+    f26(x) = circshift([1.0, 2x, 3.0], 1)[end]
+    @test autodiff(Reverse, f26, Active, Active(2.0))[1][1] == 2
+    @test autodiff(Forward, f26, Duplicated(2.0, 1.0))[1]   == 2
+    println("Done 26")
```
Comment (wsmoses) [inline test/runtests.jl:450]: remove these prints?

**Comment by wsmoses** [comment]
@jgreener64 if still worth adding, could this be rebased?

**Comment by jgreener64** [comment]
Opened https://github.com/EnzymeAD/Enzyme.jl/issues/1941.

**Comment by jgreener64** [comment]
Opened https://github.com/EnzymeAD/Enzyme.jl/issues/2102.

**Comment by jgreener64** [comment]
I think test failures are unrelated and this is ready to go.

**Pull request merged**